### ISSUE TITLE
Undefined reference errors while compiling cc1011/firmware.elf

### ISSUE THREAD

**Issue by no1knows** [JNSwanson/ESP-Home-iBoost#1]
Can't wait to try this :) I'm seeing the following errors while trying to build the firmware on my Fedora server. Any pointers gratefully received! 

```
(venv) [root@ldn ESP-Home-iBoost]# esphome run iBoost.yaml
ESPHome 2023.5.4
INFO Reading configuration iBoost.yaml...
INFO Generating C++ source...
INFO Compiling app...
Processing cc1101 (board: d1_mini; framework: arduino; platform: platformio/espressif8266@3.2.0)
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
HARDWARE: ESP8266 80MHz, 80KB RAM, 4MB Flash
Dependency Graph
|-- ESPAsyncTCP-esphome @ 1.2.3
|-- SPI @ 1.0
|-- ESPAsyncWebServer-esphome @ 2.1.0
|   |-- ESPAsyncTCP-esphome @ 1.2.3
|   |-- Hash @ 1.0
|   |-- ESP8266WiFi @ 1.0
|-- DNSServer @ 1.1.1
|-- ESP8266WiFi @ 1.0
|-- ESP8266mDNS @ 1.2
Linking .pioenvs/cc1101/firmware.elf
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy4loopEv[_ZN11iBoostBuddy4loopEv]+0x70): undefined reference to `_ZN6CC11016strobeEh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy4loopEv[_ZN11iBoostBuddy4loopEv]+0x74): undefined reference to `_ZN6CC110113writeRegisterEhh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy4loopEv[_ZN11iBoostBuddy4loopEv]+0x78): undefined reference to `_ZN6CC110118writeBurstRegisterEhPKhh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy4loopEv[_ZN11iBoostBuddy4loopEv]+0x7c): undefined reference to `_ZN6CC11019getPacketEPh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy4loopEv[_ZN11iBoostBuddy4loopEv]+0x80): undefined reference to `_ZN6CC11015crcokEv'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy4loopEv[_ZN11iBoostBuddy4loopEv]+0x84): undefined reference to `_ZN6CC11016getLQIEv'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy4loopEv[_ZN11iBoostBuddy4loopEv]+0x88): undefined reference to `_ZN6CC110110getRSSIdbmEv'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy4loopEv[_ZN11iBoostBuddy4loopEv]+0x196): undefined reference to `_ZN6CC11016strobeEh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o: in function `_ZN11iBoostBuddy4loopEv':
main.cpp:(.text._ZN11iBoostBuddy4loopEv[_ZN11iBoostBuddy4loopEv]+0x1a2): undefined reference to `_ZN6CC110113writeRegisterEhh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: main.cpp:(.text._ZN11iBoostBuddy4loopEv[_ZN11iBoostBuddy4loopEv]+0x1b1): undefined reference to `_ZN6CC110118writeBurstRegisterEhPKhh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: main.cpp:(.text._ZN11iBoostBuddy4loopEv[_ZN11iBoostBuddy4loopEv]+0x1bb): undefined reference to `_ZN6CC11016strobeEh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: main.cpp:(.text._ZN11iBoostBuddy4loopEv[_ZN11iBoostBuddy4loopEv]+0x1cf): undefined reference to `_ZN6CC11016strobeEh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: main.cpp:(.text._ZN11iBoostBuddy4loopEv[_ZN11iBoostBuddy4loopEv]+0x1e1): undefined reference to `_ZN6CC11016strobeEh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: main.cpp:(.text._ZN11iBoostBuddy4loopEv[_ZN11iBoostBuddy4loopEv]+0x1eb): undefined reference to `_ZN6CC11016strobeEh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: main.cpp:(.text._ZN11iBoostBuddy4loopEv[_ZN11iBoostBuddy4loopEv]+0x1f5): undefined reference to `_ZN6CC11016strobeEh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: main.cpp:(.text._ZN11iBoostBuddy4loopEv[_ZN11iBoostBuddy4loopEv]+0x218): undefined reference to `_ZN6CC11019getPacketEPh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: main.cpp:(.text._ZN11iBoostBuddy4loopEv[_ZN11iBoostBuddy4loopEv]+0x226): undefined reference to `_ZN6CC11015crcokEv'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: main.cpp:(.text._ZN11iBoostBuddy4loopEv[_ZN11iBoostBuddy4loopEv]+0x244): undefined reference to `_ZN6CC11016getLQIEv'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: main.cpp:(.text._ZN11iBoostBuddy4loopEv[_ZN11iBoostBuddy4loopEv]+0x29c): undefined reference to `_ZN6CC110110getRSSIdbmEv'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy5setupEv[_ZN11iBoostBuddy5setupEv]+0x28): undefined reference to `_ZN6CC11015resetEv'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy5setupEv[_ZN11iBoostBuddy5setupEv]+0x2c): undefined reference to `_ZN6CC11015beginEj'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy5setupEv[_ZN11iBoostBuddy5setupEv]+0x30): undefined reference to `_ZN6CC110113setMaxPktSizeEh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy5setupEv[_ZN11iBoostBuddy5setupEv]+0x34): undefined reference to `_ZN6CC110110setRXstateEv'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy5setupEv[_ZN11iBoostBuddy5setupEv]+0x73): undefined reference to `_ZN6CC11015resetEv'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy5setupEv[_ZN11iBoostBuddy5setupEv]+0x7e): undefined reference to `_ZN6CC11015beginEj'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy5setupEv[_ZN11iBoostBuddy5setupEv]+0x88): undefined reference to `_ZN6CC110113setMaxPktSizeEh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy5setupEv[_ZN11iBoostBuddy5setupEv]+0x94): undefined reference to `_ZN6CC110113writeRegisterEhh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy5setupEv[_ZN11iBoostBuddy5setupEv]+0xa1): undefined reference to `_ZN6CC110113writeRegisterEhh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy5setupEv[_ZN11iBoostBuddy5setupEv]+0xae): undefined reference to `_ZN6CC110113writeRegisterEhh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy5setupEv[_ZN11iBoostBuddy5setupEv]+0xba): undefined reference to `_ZN6CC110113writeRegisterEhh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy5setupEv[_ZN11iBoostBuddy5setupEv]+0xc6): undefined reference to `_ZN6CC110113writeRegisterEhh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text._ZN11iBoostBuddy5setupEv[_ZN11iBoostBuddy5setupEv]+0xd2): more undefined references to `_ZN6CC110113writeRegisterEhh' follow
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o: in function `_ZN11iBoostBuddy5setupEv':
main.cpp:(.text._ZN11iBoostBuddy5setupEv[_ZN11iBoostBuddy5setupEv]+0x22c): undefined reference to `_ZN6CC110118writeBurstRegisterEhPKhh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: main.cpp:(.text._ZN11iBoostBuddy5setupEv[_ZN11iBoostBuddy5setupEv]+0x236): undefined reference to `_ZN6CC11016strobeEh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: main.cpp:(.text._ZN11iBoostBuddy5setupEv[_ZN11iBoostBuddy5setupEv]+0x240): undefined reference to `_ZN6CC11016strobeEh'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: main.cpp:(.text._ZN11iBoostBuddy5setupEv[_ZN11iBoostBuddy5setupEv]+0x250): undefined reference to `_ZN6CC110110setRXstateEv'
/root/.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/10.3.0/../../../../xtensa-lx106-elf/bin/ld: .pioenvs/cc1101/src/main.cpp.o:(.text.startup._GLOBAL__sub_I_logger_logger+0x0): undefined reference to `_ZN6CC1101C1EhhR8SPIClass'
collect2: error: ld returned 1 exit status
*** [.pioenvs/cc1101/firmware.elf] Error 1
================================================================================================================================================================================ [FAILED] Took 3.11 seconds ================================================================================================================================================================================

```

**Comment by no1knows** [comment]
Resolved with:
`cp CC1101_RFx.* .esphome/build/cc1101/src/`

Now I'm getting the following in the esphome console after flashing:
```
[20:23:15]SPI OK
[20:23:15][W][iBoost:111]: SPI OK
[20:23:17]
[20:23:17]--------------- CUT HERE FOR EXCEPTION DECODER ---------------
[20:23:17]
[20:23:17]Soft WDT reset
[20:23:17]
[20:23:17]>>>stack>>>
[20:23:17]
[20:23:17]ctx: cont
[20:23:17]sp: 3ffffc70 end: 3fffffd0 offset: 01a0
[20:23:17]3ffffe10:  00000000 00000001 0000000f 40100674  
[20:23:17]3ffffe20:  33c134e0 00000000 0000000f 40100674  
[20:23:17]3ffffe30:  3ffffe60 0000003d 3ffefd1c 4020cba2  
[20:23:17]3ffffe40:  3ffffe60 33c134e0 3ffefd1c 4020cc70  
[20:23:17]3ffffe50:  3ffe9024 33c134e0 3ffefd1c 4020cde1  
[20:23:17]3ffffe60:  3ffe9024 33c134e0 3ffefd1c 4020cf71  
[20:23:17]3ffffe70:  00000000 3ffefeb0 3ffefd1c 4021d57e  
[20:23:17]3ffffe80:  3fff150c 00000009 3ffefc14 4021ba61  
[20:23:17]3ffffe90:  00000006 00000006 3fff1c14 40100d82  
[20:23:17]3ffffea0:  00000028 00000014 00000005 4023aa3c  
[20:23:17]3ffffeb0:  3fff1954 6d5f7473 3fffff80 3fff111c  
[20:23:17]3ffffec0:  3ffefc14 3ffefd28 3fffff28 4021e86f  
[20:23:17]3ffffed0:  4023146c 00000000 7fc00000 00000000  
[20:23:17]3ffffee0:  3fff13a4 3fff13c0 3fff13c0 c75f640a  
[20:23:17]3ffffef0:  3ffffef8 0000000a 706d6554 34333231  
[20:23:17]3fffff00:  00003635 00000000 00000000 00000000  
[20:23:17]3fffff10:  00000000 00000000 00000000 00000000  
[20:23:17]3fffff20:  00000000 00000000 3fff19c4 3fff19c8  
[20:23:17]3fffff30:  3fff19c8 00000000 ebadba69 ee1f95a3  
[20:23:17]3fffff40:  873a893e c75f640a 0c004487 1aefeae3  
[20:23:17]3fffff50:  b3021969 6f5a9ea8 ebadba69 ee1f95a3  
[20:23:17]3fffff60:  873a893e c75f640a 3fffff80 00000001  
[20:23:17]3fffff70:  00000000 3fffff68 00000001 00000000  
[20:23:17]3fffff80:  3fff19a4 feefeffe feefeffe feefeffe  
[20:23:17]3fffff90:  3fff190c 3fff1954 feefeffe feefeffe  
[20:23:17]3fffffa0:  feefeffe feefeffe feefeffe 3fff0078  
[20:23:17]3fffffb0:  3fffdad0 00000000 3fff0064 4022c7c8  
[20:23:17]3fffffc0:  feefeffe feefeffe 3ffe8618 4010043d  
[20:23:17]<<<stack<<<
[20:23:17]
[20:23:17]--------------- CUT HERE FOR EXCEPTION DECODER ---------------
[20:23:17]
[20:23:17] ets Jan  8 2013,rst cause:2, boot mode:(3,6)
[20:23:17]
[20:23:17]load 0x4010f000, len 3460, room 16 
[20:23:17]tail 4
[20:23:17]chksum 0xcc
[20:23:17]load 0x3fff20b8, len 40, room 4 
[20:23:17]tail 4
[20:23:17]chksum 0xc9
[20:23:17]csum 0xc9
[20:23:17]v0007b6e0
[20:23:17]~ld
```

**Comment by JNSwanson** [comment]
Have you got the radio module attached?  It is the WDT resetting probably when trying to set up the SPI or radio.

**Comment by no1knows** [comment]
I do... and have double checked the wiring (photo attached). I haven't tested my CC1101 module, and the build quality doesn't seem great so it could be a dud. I'll do some testing with some basic CC1101 example scripts and see how I get on.
![image](https://github.com/JNSwanson/ESP-Home-iBoost/assets/263718/d22340e5-66b7-48cc-bacf-0e1b7640de75)

**Comment by no1knows** [comment]
Seems to be working fine with some simple Arduino example sketches. Definitely no need to connect GDO0 and GDO2?

**Comment by JNSwanson** [comment]
I don’t think a buddy was necessary for the code to work, but it was very useful in decoding the protocol. You don’t need to pair it as it just uses the strongest signal. You should be seeing packets from the sender at least, so I suspect the rx is a little off frequency. The one I tried was out by 75kHz. It will only work with units that support the Buddy ie the iBoost+ and not the original iBoost.

**Comment by JNSwanson** [comment]
Btw I used sdrsharp to check the frequency offset. You would need to modify the code to force a transmission so you can see the difference

**Comment by no1knows** [comment]
Thanks for the pointers. I may be out of my depth somewhat here - I wonder if there's software-defined radio that's more accurate out of the box. I'll see how I get on with some of your other suggested frequencies.

**Comment by JNSwanson** [comment]
The middle of the table vertically is the nominal centre frequency and the entries go up and down by 25kHz. You will need to edit the 3 lines where the frequency is defined and see if you start receiving packets. I would try +/- 50kHz then +/- 100kHz.  Given the modules seem to be out of tolerance I might try to automate the process. It may help if you are close to the sender unit.

**Comment by no1knows** [comment]
Got it! Just realised that the 21656A in the table corresponds to:

```
radio.writeRegister(CC1101_FREQ2, **0x21**); // 868.300MHz  (868300000 <<16)/26000000
radio.writeRegister(CC1101_FREQ1, **0x65**);
radio.writeRegister(CC1101_FREQ0, **0x6a**);
```

Will try some others out. An auto solution would also be great :) 

I am slightly concerned that without pairing a Buddy the unit won't be transmitting anything, but let's see.

**Comment by JNSwanson** [comment]
The code emulates what the Buddy does, so I don’t think you will need one.

**Comment by JNSwanson** [comment]
It doesn’t start transmitting until it has received a packet from the sensor or a Buddy

**Comment by JNSwanson** [comment]
In any case you should see the received packets in the log.

**Comment by no1knows** [comment]
Sadly no luck with +/- 50Khz and +/- 100khz - no sign of any packets in the log. Will try again closer to the sender and, in the meantime, may order the Epson xtal!

**Comment by JNSwanson** [comment]
Maybe try the 25kHz spacing.  I noticed from you photo that D8 isn't connected.  I assume you have fixed this?

**Comment by JNSwanson** [comment]
I tried it with the Buddy disconnected and it works ok.  You do need to be in range of the sender though.

**Comment by no1knows** [comment]
Ah, interesting. The sender is a long way away, which I suppose means I need to get the frequency spot on. I just replaced the crystal with the Epson X1E0000210666 from Farnell (2471832), reset the frequency settings in iBoost.h and... hey presto:

```
[22:08:38]Got packet,27,f2,01,00,0b,08,0b,12,25,2b,20,00,00,00,00,00,00,00,00,00,00,08,0e,0a,09,0b,0d,22,2d,21,05,00,00,00,00,00,00,00,00,00,00,0b,0f,03,,len=44 Signal=-91 LQI=52
[22:08:38]Updated the address to:27,f2
[22:08:39]Sending request
```
All data is now showing in HA, and the boost/cancel buttons are working too. Awesome! Thanks so much for creating this, and all your help troubleshooting!

**Comment by maksyms** [comment]
I just wanted to leave this here as well: I experienced the same issue as @no1knows, but I followed the advise to resolder the crystal - and all is working right away! Thank you!

**Comment by tobes49** [comment]
JNSwanson Thank you for all your hard work. 

After trying to get it working and wouldn't I managed to resolder the crystal (using a soldering iron, took 3 attempts) but now all working and love having my iboost data in Home Assistant

Thank you -  PS have you got a buymeacoffee or something like that?

**Comment by no1knows** [comment]
Since this issue is still open - QQ - has anyone managed to use the esphome addon to install/update the firmware? I just tried, and got the usual CC1101_RFx-related errors. Does anyone know where you're supposed to copy the .cpp and .h files so that the esphome addon can see them when compiling? (Presumably somewhere in a docker container?). For now I've compiled the latest version back on my server (rather than RPi) manually using esphome cmdline, but would be great if we could update without doing that. Cheers!
